src/core/qgsvectorlayer.cpp

```diff
@@ -2243,10 +2248,34 @@ bool QgsVectorLayer::changeGeometry( QgsFeatureId fid, const QgsGeometry &geom )
 
 bool QgsVectorLayer::changeAttributeValue( QgsFeatureId fid, int field, const QVariant &newValue, const QVariant &oldValue )
 {
-  if ( !mEditBuffer || !mDataProvider )
-    return false;
+  if ( fields().fieldOrigin( field ) == QgsFields::OriginJoin )
+  {
+    int srcFieldIndex;
+    const QgsVectorLayerJoinInfo *info = mJoinBuffer->joinForFieldIndex( field, fields(), srcFieldIndex );
+    if ( info && info->joinLayer() && info->isEditable() )
+    {
+      QgsFeature feature = getFeature( fid );
+
+      if ( !feature.isValid() )
+        return false;
+
+      QgsFeature joinFeature = mJoinBuffer->joinedFeatureOf( info, feature );
 
-  return mEditBuffer->changeAttributeValue( fid, field, newValue, oldValue );
+      if ( joinFeature.isValid() )
+        return info->joinLayer()->changeAttributeValue( joinFeature.id(), srcFieldIndex, newValue, oldValue );
```

Comment:
That's fair enough - but in this case we need to give helpful feedback explaining why the edit failed. Silently failing isn't going to do us any favours with our users ;)

I'm also curious how this approach would work with the use case of joined columns for label data defined x/y (I can't find the relevant discussion now - but somewhere I remember a lengthy discussion about joins and automatically allowing movement of labels... any ideas where this is?).  Wouldn't this use case require automatic switching on of edit mode for the joined table containing x/y columns?